src/coreclr/jit/importer.cpp

```diff
@@ -12560,7 +12557,8 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
         //
         // use the special helper that skips the cases checked by our inlined cast
         //
-        const CorInfoHelpFunc specialHelper = CORINFO_HELP_CHKCASTCLASS_SPECIAL;
+        const CorInfoHelpFunc specialHelper =
+            (helper == CORINFO_HELP_CHKCASTCLASS) ? CORINFO_HELP_CHKCASTCLASS_SPECIAL : helper;
```

Comment:
What other `helper` we expect for `isCastClass == true`?

Reply:
@kunalspathak only `CORINFO_HELP_CHKCASTCLASS` or `CORINFO_HELP_CHKCASTARRAY`

Reply:
Can we  `assert(helper == CORINFO_HELP_CHKCASTARRAY)`?

Reply:
Hm. actually no, it also can be `CORINFO_HELP_CHKCASTINTERFACE` but only in case of profiled `castclass` (turned off by default) let me fix

Reply:
@kunalspathak added

Reply:
`CORINFO_HELP_CHKCASTINTERFACE` wasn't handled previously (sounds like a bug) but it was only ever used with `DOTNET_JitProfileCasts=1` which is 0 by default

Reply:
@jkotas to make sure we're on the same page, so previously it looked like this:

![image](https://user-images.githubusercontent.com/523221/191311518-13610b65-3ac5-4f84-92af-439890e61ef3.png)

while this PR changes the fallback one to `CORINFO_HELP_CHKCASTARRAY` that sounds more appropriate, no?

Reply:
Yes, the fallback to `CORINFO_HELP_CHKCASTINTERFACE` is more appropriate in this case.

Also, revert the change that was done in https://github.com/dotnet/runtime/pull/69869 in ChkCastClassSpecial. The whole purpose of ChkCastClassSpecial is to avoid this extra check that #69869 added. (IIRC, avoiding this extra check was important for some microbenchmarks. It is interesting that we have not noticed it.)

Reply:
Fixed, thanks for noticing it!
`ChkCastClass won't do the right thing - it is a functional bug.`
From my understanding it will just forward check to FCall ending up in ObjIsInstanceOfCore where interfaces are handled, so it's not a bug but rather a penalty for mispredicted fast path

Reply:
Yes, you are right. We are lucky that it happens to work.

Reply:
@jkotas actually it looks like the check you asked me to delete is still needed, consider this example:
```csharp
class ClassA { }
class ClassB : ClassA { }

internal class Program 
{

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static void Main()
    {
        for (int i = 0; i < 100; i++)
        {
            CastToClassA(new ClassB());
            Thread.Sleep(10);
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static ClassA CastToClassA(object o) => (ClassA)o;
}
```
Current Tier1 codegen for `CastToClassA` is:
```asm
; Assembly listing for method CastToClassA
G_M33135_IG01:              
       4883EC28             sub      rsp, 40
G_M33135_IG02:              
       48894C2430           mov      gword ptr [rsp+30H], rcx
       488BC1               mov      rax, rcx
       4885C0               test     rax, rax
       741D                 je       SHORT G_M33135_IG05
G_M33135_IG03:              
       48BA18537F78FE7F0000 mov      rdx, 0x7FFE787F5318      ; ClassA
       483910               cmp      qword ptr [rax], rdx
       740E                 je       SHORT G_M33135_IG05
G_M33135_IG04:              
       488BCA               mov      rcx, rdx
       488B542430           mov      rdx, gword ptr [rsp+30H]
       FF152A75FBFF         call     [CORINFO_HELP_CHKCASTCLASS_SPECIAL]
G_M33135_IG05:              
       90                   nop      
G_M33135_IG06:              
       4883C428             add      rsp, 40
       C3                   ret
```
So jit inlined check for `ClassA` so it doesn't need to check for it inside the _SPECIAL helper. But as you can see from the call-site the check for `ClassA` fast path is not valid for given app in terms of performance. So with Dynamic PGO and `DOTNET_JitProfileCasts=1` I get this:
```asm
; Assembly listing for method CastToClassA
G_M33135_IG01:
       4883EC28             sub      rsp, 40
G_M33135_IG02:              
       488BC1               mov      rax, rcx
       4885C0               test     rax, rax
       7422                 je       SHORT G_M33135_IG05
G_M33135_IG03:              
       48BA484F8078FE7F0000 mov      rdx, 0x7FFE78804F48      ; ClassB
       483910               cmp      qword ptr [rax], rdx
       7413                 je       SHORT G_M33135_IG05
G_M33135_IG04:              
       488BD1               mov      rdx, rcx
       48B9E04D8078FE7F0000 mov      rcx, 0x7FFE78804DE0      ; ClassA
       FF157A75FBFF         call     [CORINFO_HELP_CHKCASTCLASS_SPECIAL]
G_M33135_IG05:              
       90                   nop      
G_M33135_IG06:              
       4883C428             add      rsp, 40
       C3                   ret
```
Here we check for `ClassB` for fast path but we'll need the check for itself in the helper. Here is the codegen diff between two (PGO is on the right):

Reply:
As an optimization, we can introduce `CORINFO_HELP_CHKCASTCLASS_SPECIAL_SLOW` with the "am I the same type?" check

Reply:
Alternatively, we can check for both in JIT inlined

Reply:
The JIT should be using the regular `CORINFO_HELP_CHKCASTCLASS` when it decides to inline a check for a type from the middle of the type hierarchy.

> CORINFO_HELP_CHKCASTCLASS_SPECIAL_SLOW with the "am I the same type?" check

It is exactly what `CORINFO_HELP_CHKCASTCLASS` does. We do not need a new helper.

Reply:
Ah, indeed 🙂